llvm/lib/CodeGen/MachineBlockPlacement.cpp

```diff
@@ -3158,9 +3158,9 @@ bool MachineBlockPlacement::maybeTailDuplicateBlock(
 
         // Handle the filter set
         if (BlockFilter) {
-          BlockFilter->remove(RemBB);
           if (*PrevUnplacedBlockInFilterIt == RemBB)
             PrevUnplacedBlockInFilterIt++;
+          BlockFilter->remove(RemBB);
```

Comment:
With this modification to BlockFilter, PrevUnplacedBlockInFilterIt becomes invalidated.

Reply:
It looks like PrevUnplacedBlockInFilterIt does not need to be modified at all because it does not increment automatically in buildChain()

Reply:
PrevUnplacedBlockInFilterIt becomes invalidated after this statement, even you don't increment it immediately, you still want to use it in getFirstUnplacedBlock later.

Reply:
BlockFilter is a SmallSetVector (which is backed by a small vector) and iterators are tied to the underlying vector (instead of the set), and removing an element shifts following elements one position forward without reallocation so the iterator should be valid (and pointing to the following element if remove happens at or before PrevUnplacedBlockInFilterIt)

Reply:
Code is updated.  On a side note why couldn't SmallVector guarantee iterator validity after element removal, given that LLVM is in control of its implementation?